### PR TITLE
[torchgen] accept scalars for unary `SymInt` arrays

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #99921

Fixes https://github.com/pytorch/pytorch/issues/99907

### CODE REVIEW

File: torchgen/api/python.py [hunk 1]
```diff
@@ -951,6 +951,9 @@ def argument_type_str_pyi(t: Type) -> str:
             )
         elif str(t.elem) == "float":
             ret = "Sequence[_float]"
+        elif str(t.elem) == "SymInt" and t.size == 1:
+            elem = argument_type_str_pyi(t.elem)
+            ret = f"Union[{elem}, Sequence[{elem}]]"
```
Comment (nkaretnikov) [inline torchgen/api/python.py:956]: Non-list SymInts are handled like this:

```py
        if t.name == BaseTy.SymInt:
            ret = "Union[_int, SymInt]"
```

We have this rule for ints that are in lists:

```py
    elif isinstance(t, ListType):
        if str(t.elem) == "int":
            ret = "Union[_int, _size]" if t.size is not None else "_size"
```

I don't quite understand it, is it even correct? So I went with a more conservative rule for SymInts.

New signature of roll. It has two Unions, but I think it's better than inlining and hardcoding in torchgen.

```py
def roll(input: Tensor, shifts: Union[Union[_int, SymInt], Sequence[Union[_int, SymInt]]], dims: Union[_int, _size] = ()) -> Tensor: ...
```

**Comment by nkaretnikov** [comment]
@pmeier Please verify that this solves https://github.com/pytorch/vision/issues/7522. Not roll-specific, we just have a special rule for ints, but we didn't have one for SymInts.

@ezyang PTAL. Can we do better? Note: this still doesn't fix https://github.com/pytorch/pytorch/pull/99114#discussion_r1166169890 as that's separate.

File: torchgen/api/python.py [hunk 2]
```diff
@@ -951,6 +951,9 @@ def argument_type_str_pyi(t: Type) -> str:
             )
         elif str(t.elem) == "float":
             ret = "Sequence[_float]"
+        elif str(t.elem) == "SymInt" and t.size == 1:
```
Comment (malfet) [inline torchgen/api/python.py:954]: Why limit to size of 1? Also see https://github.com/pytorch/pytorch/issues/99907 (reassigned it to you)

**Review by malfet — APPROVED** [review]

File: torchgen/api/python.py (hunk 2 above)
Comment (nkaretnikov) [reply torchgen/api/python.py:954]: Maybe I'm misreading this, but when the size is not 1, it should just be a sequence as before, no? So will be handled by the `else` block.

Comment (malfet) [reply torchgen/api/python.py:954]: Would single int be broadcastable to all dimension? For consistency with old behavior, `_int` was added as an acceptable type if `t.size is not None`, but here you are restricting it to `t.size == 1`. Also, you technically need to add `torch.Size` to the list

**Pull request closed without merging**